src/or/onion.c

```diff
@@ -1339,3 +1342,466 @@ extended_cell_format(uint8_t *command_out, uint16_t *len_out,
+  }
+
+  if (len == 0) {
+    log_info(LD_OR, "Received a strange CREATE(D)2V cell whose reported "
+             "length was 0 bytes.");
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Parse the CREATE2V payload at <b>p</b>, which could be up to <b>p_len</b>
+ * bytes long, and use it to fill the fields of <b>cell_out->body</b>.
+ *
+ * Note that part of the body of an EXTEND2V cell is a CREATE2V payload, so
+ * this function is also used for parsing those.
+ *
+ * Returns true if parsing was successful, and false otherwise.
+ */
+static bool
+parse_create2v_payload(create2v_cell_t *cell_out,
+                       const uint8_t *p,
+                       const size_t p_len)
+{
+  create2v_cell_body_t *body;
+
+  /* Check the length before even attempting to parse it. */
+  if (!create2v_cell_check_encoded_length(p_len)) {
+    return false;
+  }
+  create2v_cell_body_parse(&body, p, p_len);
+  cell_out->body = body;
+
+  return create2v_cell_check(cell_out, true);
+}
+
+/**
+ * Parse a CREATE2V cell from <b>cell_in</b> into <b>cell_out</b>.
+ *
+ * Returns true if parsing was successful, and false otherwise.
+ */
+bool
+create2v_cell_parse(create2v_cell_t *cell_out, const var_cell_t *cell_in)
+{
+  switch (cell_in->headers.command) {
+  case CELL_CREATE2V:
+    return parse_create2v_payload(cell_out,
+                                  cell_in->payload,
+                                  cell_in->payload_len);
+  default:
+    return false;
+  }
+}
+
+/**
+ * Run various checks to ensure that a create2v_cell_t, <b>cell</b>, is
+ * well-formed.
+ *
+ * If <b>unknown_ok</b> is true, then handshake types we don't know about yet
+ * are okay.
+ *
+ * In the future, checks such as verifying that the handshake data (for a
+ * particular handshake) should go here.
+ *
+ * Returns true if the <b>cell</b> is okay, and false otherwise.
+ */
+bool
+create2v_cell_check(const create2v_cell_t *cell, bool unknown_ok)
+{
+  const char *not_okay;
+
+  not_okay = create2v_cell_body_check(cell->body);
+
+  if (not_okay) {
+    log_warn(LD_BUG, "Unable to parse a CREATE2V cell: %s", not_okay);
+    return false;
+  }
+
+  switch (create2v_cell_body_get_htype(cell->body)) {
+  case ONION_HANDSHAKE_TYPE_NTOR2_NULL:
+    if (create2v_cell_body_get_hlen(cell->body) >
+        ONION_HANDSHAKE_TYPE_NTOR2_NULL_MAXLEN) {
+      return false;
+    }
+    break;
+  default:
+    if (! unknown_ok)
+      return false;
+  }
+  return true;
+}
+
+/**
+ * Fill <b>cell_out</b> with up to <b>payload_len</b> bytes of a correctly
+ * formatted version of the CREATE2V cell in <b>cell_in</b>.
+ *
+ * If <b>relayed</b> is true, then this is a cell not originating from us.
+ *
+ * Returns true if the formatting was successful, and false otherwise.
+ */
+static bool
+create2v_cell_format_impl(var_cell_t *cell_out,
+                          const size_t payload_len,
+                          const create2v_cell_t *cell_in,
+                          const bool relayed)
+{
```

Comment:
Would it be cleaner for this function to allocate and return the var_cell_t ?  That way you don't need to worry about heap overruns if payload_len is not the allocated payload length of cell_out.